libmeepgeom/dft-fields.cpp

```diff
@@ -0,0 +1,116 @@
+/* unit test for dft_fields and output_dft functionality       */
+/***************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <complex>
+#include <vector>
+
+#include "meep.hpp"
+
+#include "ctl-math.h"
+#include "ctlgeom.h"
+
+#include "meepgeom.hpp"
+
+using namespace meep;
+
+typedef std::complex<double> cdouble;
+
+vector3 v3(double x, double y=0.0, double z=0.0)
+{
+  vector3 v;
+  v.x=x; v.y=y; v.z=z;
+  return v;
+}
+
+/***************************************************************/
+/* dummy material function needed to pass to structure( )      */
+/* constructor as a placeholder before we can call             */
+/* set_materials_from_geometry                                 */
+/***************************************************************/
+double dummy_eps(const vec &) { return 1.0; }
+
+/***************************************************************/
+/***************************************************************/
+/***************************************************************/
+void Run(bool Pulse)
+{
+  /***************************************************************/
+  /* initialize geometry                                         */
+  /***************************************************************/
+  double n=3.4;     // index of waveguide
+  double w=1.0;     // width of waveguide
+  double r=1.0;     // inner radius of ring
+  double pad=4;     // padding between waveguide and edge of PML
+  double dpml=2;    // thickness of PML
+
+  double resolution = 10.0;
+
+  double sxy        = 2.0*(r+w+pad+dpml);  // cell size
+  geometry_lattice.size.x=sxy;
+  geometry_lattice.size.y=sxy;
+  geometry_lattice.size.z=0.0;
+  grid_volume gv = voltwo(sxy, sxy, resolution);
+  gv.center_origin();
+  symmetry sym=identity(); // mirror(Y, gv);
+  structure the_structure(gv, dummy_eps, pml(dpml), sym);
+
+  /***************************************************************/
+  /* add objects                                                 */
+  /***************************************************************/
+  meep_geom::material_type dielectric = meep_geom::make_dielectric(n*n);
+  geometric_object objects[2];
+  vector3 v3zero = {0.0, 0.0, 0.0};
+  vector3 zaxis  = {0.0, 0.0, 1.0};
+  objects[0] = make_cylinder(dielectric, v3zero, r+w, ENORMOUS, zaxis);
+  objects[1] = make_cylinder(meep_geom::vacuum,  v3zero, r, ENORMOUS, zaxis);
+  geometric_object_list g={ 2, objects };
+  meep_geom::set_materials_from_geometry(&the_structure, g);
+  fields f(&the_structure);
+  f.step(); // single timestep to trigger internal initialization
+  f.output_hdf5(Dielectric, f.v);
+
+  /***************************************************************/
+  /***************************************************************/
+  /***************************************************************/
+  double fcen = 0.15;  // ; pulse center frequency
```

Comment:
Note that, as shown in the [tutorial example](http://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#modes-of-a-ring-resonator), the fundamental mode (with three-fold symmetry) for this structure is at a frequency of 0.118. The next higher-order modes are at 0.147 and 0.175. The convergence properties of the frequency-domain solver may be improved by setting the CW source frequency to one of these modes.